### PR TITLE
Validate configuration at the end of parsing

### PR DESCRIPTION
**Pull request by luhring**
This ensures we fail fast when detect that a user-supplied value in the build configuration is not valid.

This helps guard against package names like `boost-"atomic"`, which we saw recently (see https://github.com/wolfi-dev/os/pull/2269).

**Note:** For names of packages and subpackages, this implementation requires the following:
- all letters are lowecase
- must start with letter
- the only valid characters are letters, numbers, periods (`.`), dashes (`-`), and underscores (`_`)

Underscores in package names are very rare, but they do exist in Wolfi right now. If we wanted, we could disallow underscores and update any offending package names accordingly:

- `libnetfilter_conntrack`
- `libnetfilter_cthelper`
- `libnetfilter_cttimeout`
- `libnetfilter_queue`
- `ruby3.2-http_parser.rb`

### CODE REVIEW

File: pkg/build/build.go [hunk 1]
```diff
@@ -1170,9 +1171,44 @@ func ParseConfiguration(configurationFilePath string, opts ...ConfigurationParsi
...
 
+	// Finally, validate the configuration we ended up with before returning it for use downstream.
+	if err = cfg.validate(); err != nil {
+		return nil, fmt.Errorf("validating configuration: %w", err)
+	}
+
 	return &cfg, nil
 }
 
+type ErrInvalidConfiguration struct {
```
[the remark this answers is not in the document]
Comment (kaniini) [reply pkg/build/build.go:1182]: Yeah, I too am curious about why we use a struct here

[the remark this answers is not in the document]
Comment (luhring) [reply pkg/build/build.go:1182]: (I'm referring to this, in case that wasn't clear: https://go.dev/blog/go1.13-errors)

**Comment by kaniini** [comment]
Let me check our YAMLs against this branch before merging this time :)

**Comment by kaniini** [comment]
```
db.yaml
Error: validating configuration: build configuration is invalid: subpackage name "db-c++" (subpackages index: 3) must match regex "^[a-z][a-z0-9_.-]*$"
2023/05/24 09:49:20 error during command execution: validating configuration: build configuration is invalid: subpackage name "db-c++" (subpackages index: 3) must match regex "^[a-z][a-z0-9_.-]*$"
```

I think `db-c++` is a valid subpackage name, as it is the one used in Alpine.

**Comment by kaniini** [comment]
The regexp would also block something like `389-directory-server`, which is a real package in some other distros.

**Comment by luhring** [comment]
@kaniini Added support for `+` and uppercase letters, and for numbers to be able to start the package name.

**Review by kaniini — APPROVED** [review]

**Pull request merged**